flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java

```diff
@@ -96,28 +98,70 @@ public JdbcDynamicTableSource(
     public LookupRuntimeProvider getLookupRuntimeProvider(LookupContext context) {
         // JDBC only support non-nested look up keys
         String[] keyNames = new String[context.getKeys().length];
-        for (int i = 0; i < keyNames.length; i++) {
+
+        for (int i = 0; i < context.getKeys().length; i++) {
             int[] innerKeyArr = context.getKeys()[i];
             Preconditions.checkArgument(
                     innerKeyArr.length == 1, "JDBC only support non-nested look up keys");
             keyNames[i] = DataType.getFieldNames(physicalRowDataType).get(innerKeyArr[0]);
         }
+
         final RowType rowType = (RowType) physicalRowDataType.getLogicalType();
+
+        String[] conditions = null;
+
+        if (this.resolvedPredicates != null) {
+            conditions = new String[this.resolvedPredicates.size()];
+            for (int i = 0; i < this.resolvedPredicates.size(); i++) {
+                String resolvedPredicate = this.resolvedPredicates.get(i);
+                String param = this.pushdownParams[i].toString();
+                /*
+                 * This replace seems like it should be using a Flink class to resolve the parameter. It does not
+                 * effect the dialects as the placeholder comes from JdbcFilterPushdownPreparedStatementVisitor.
+                 *
+                 * Here is what has been considered as alternatives.
+                 *
+                 * We cannot use the way this is done in getScanRuntimeProvider, as the index we have is the index
+                 * into the filters, but it needs the index into the fields. For example one lookup key and one filter
+                 * would both have an index of 0, which the subsequent code would incorrectly resolve to the first
+                 * field.
+                 * We cannot use the PreparedStatement as we have not got access to the statement here.
+                 * We cannot use ParameterizedPredicate as it takes the filter expression as input (e.g EQUALS(...)
+                 * not the form we have here an example would be ('field1'= ?).
+                 */
+                conditions[i] = resolvePredicateParam(resolvedPredicate, param);
```

Comment:
I tried to debug and was not able to find any test which tests this piece of code
Did I miss anything?

Reply:
Yes the tests are too big of a diff in the default view in git - if you expand out the changes you will see that flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceITCase.java drives this method.

Reply:
ok I see, thanks
then it seems it doesn't take into account key names and as a result returns wrong value for complex join conditions like e.g.
```sql
SELECT * FROM a left join c FOR SYSTEM_TIME AS OF a.proctime on (c.type = 0 or c.ip = '??') and a.ip = c.ip
```

we need to use `keyNames` info as well

Reply:
@snuyanzin great spot - I will look into this

Reply:
@snuyanzin I am reading JdbcFilterPushdownPreparedStatementVisitor and it looks like it only attempts to pushdown simple join expressions; in which case this code is ok (as that was my assumption). I will test to confirm my understanding is correct.

Reply:
I just tested it and it fails. I need to relook at the fix so that is can handle complex joins like the one you mention. 
Great catch :-)

Reply:
I did a more sophisticated query with more than one resolvedPredicate and multiple pushdownParams in one of them.   
![image](https://github.com/apache/flink-connector-jdbc/assets/39792797/8a360200-61cc-4217-ab21-13eacf7e1b1e)

I am thinking I need to replace the ? characters with the pushdownParams in order. @snuyanzin is this what you meant by "account key names"?

Reply:
thanks for looking in this
yep, you are right since `keyNames` has info about order and we can reuse it

Reply:
fixed this